### PR TITLE
Add user data hash to JobBatch

### PR DESCRIPTION
**Pull request by vadim-ex**

### CODE REVIEW

**Comment by myronmarston** [comment]
The interface here feels kinda un-rubyish -- in general, "get" and "set" in ruby method names feel un-idiomatic given that ruby supports `foo`/`foo=` and `[]`/`[]=` as method names, which allows for assignment syntax.

On top of that, it's unclear to me what the advantage of defining custom `user_data_get`, `user_data_set` and `user_data_keys` methods are, rather than just considering `user_data` a public API, which returns a [Redis::HashKey](https://github.com/nateware/redis-objects/blob/v1.1.0/lib/redis/hash_key.rb), which supports these 3 operations plus many more with a familiar hash-like interface. So, for example, if I want to check if a key is in the hash, `Redish::HashKey` gives me an O(1) `has_key?` method, whereas using `user_data_keys.include?(key)` is two O(N) operations (one in redis, one in ruby).  Or if I want to fetch a value with a default fallback (a common thing with hashes -- I use `Hash#fetch` very frequently), `Redish::HashKey` provides that but your interface does not.

Alternately, I think it might be worth considering storing user data in the `meta` hash, rather than using a separate hash.  The nice thing about re-using the meta hash is that when troubleshooting, I frequently use `job_batch.meta.all` to print out all the job batch's metadata, in order to learn whatever is available about the batch's history.  Having a separate `user_data` hash forces me to look at two things to get all that data.  OTOH, if we expect `user_data` to grow quite large, I can definitely see the argument in favor of keeping it separate, although I would be surprised to see it grow large.  If we do decide to keep it in the meta hash, I'd recommend something like the following:

``` ruby
def user_data
  user_data_json = meta["user_data"]
  (user_data_json && JSON.parse(user_data_json)) || {}
end

def user_data=(new_user_data)
  meta["user_data"] = JSON.generate(new_user_data)
end
```

Under this scheme, the user data would be nested under the `meta` hash in a `user_data` key in order to prevent naming clashes.  If we went with this API, mutations made to the object returned from `user_data` would not automatically be persisted to redis like they would if we expose a `Redish::HashKey`, but I think that's OK -- instead this supports explicit assignment to update the user data when you want it persisted.

Thoughts?

**Comment by koos42** [comment]
I disagree, if `user_data` were absorbed into the `meta` hash,  `Redish::HashKey` it would make it too easy to have race conditions when updating.

I imagine that a potential application that has multiple jobs updating separate keys in the user_data could easily have jobs negate each others' changes by parsing the stored JSON, modifying it, then saving it back around the same time.

The only fix for that would be managing the updates within Lua script, which sounds painful.

**Comment by koos42** [comment]
Do `Redish::HashKey`s automagically save on updating the value of their keys?

**Comment by myronmarston** [comment]
You make a great point about potential race conditions. I didn't think of that.

All of the objects provided by redis-objects track only two bits of state: a redis connection and a redid key. All operations on the objects internally delegate to redis calls. Thus all mutation methods provided by these objects update redis since they don't keep any local copies of the data.

**Comment by vadim-moz** [comment]
Myron, thank you for the long and detailed comment. We would be really glad to discuss the change in more details before submitting PR, but we appreciate you were and are busy. The good part, it is not late to change now.

Let me clarify intentions and design decisions. First, the concerns design tried to address:
* minimize potential of race condition;
* draw explicit line between Plines handled, and Plines contained data;
* allow independent use of container by subsystems.

From this concerns came design decisions:
* The container should be redis::hash. The json-packed key in meta would produce race conditions on update and inter-subsystem dependencies.
* Currently defined methods operate on group of keys. Thus operation is pseudo-transaction. In fact, operations atomicity was one of driving forces. (It also dictated choice of methods' names for a degree: [] is single-key method.)
* Implement only minimal set of operations. It is easy to give, it is difficult to take.

The explicit exposure of user_data is a possibility. The strong part of it is simplicity. There are some disadvantages:
* abstraction leak. Implementation of user data container as a redis::hash is accidental.
* tight coupling and [law of Demeter](http://en.wikipedia.org/wiki/Law_of_Demeter) violation. The law is not infallible, but its violation is a smell.
In fact, the api veil would allow us to change implementation from redis::hash to meta["user_data"] any moment we'd consider it beneficial.

Still, we are ready to switch to exposed user_data.

**Comment by myronmarston** [comment]
Yep, it's all tradeoffs.  There isn't an obvious winner; it comes down to what pros/cons you decide are more valuable/important and that's ultimately subjective.  I think I have a slight preference for making `user_data` public and removing the accessor methods you added but it is very slight, and I trust your judgement -- particularly since you've thought through this so intentionally. 

If you stick with the accessor methods, my one request would be to change their names:

* `user_data_get` -> `get_user_data`
* `user_data_set` -> `set_user_data`

IMO, `verb_noun` reads much more naturally than `noun_verb` and is easier to remember.  Consider that it's `Hash#has_key?`, not `Hash#key_has?`, `Module#define_method`, not `Module#method_define`, etc.

(Although admittedly, Ruby also has `instance_variable_get` and `instance_variable_set`, but those have always tripped me up).

File: spec/unit/plines/job_batch_spec.rb [hunk 1]
```diff
@@ -1097,5 +1097,52 @@ def cancel(jb = batch, *args)
         end
       end
     end
+
+    context 'userw data handling' do
```
Comment (myronmarston) [inline spec/unit/plines/job_batch_spec.rb]: s/userw/user/

Comment (koos42) [inline spec/unit/plines/job_batch_spec.rb]: s/w//

**Comment by myronmarston** [comment]
And one other suggestion that just popped in my head: having to pass keys in an array to `get_user_data` seems suboptimal...I prefer to use ruby's splat feature for this:

``` ruby 
def get_user_data(*keys)
  # keys is an array
end

get_user_data('key1') # get just 'key1'
get_user_data('key1', 'key2') # get both keys
get_user_data # get all data (the `keys` arg will be an empty array)
```

The one "gotcha" is that when you have an array you are trying to pass you have to splat it out:

```
keys = ['key1', 'key2']
get_user_data(*keys)
```

I think that's a reasonable tradeoff, particularly if you expect the more common case to be callers passing literal keys.

**Comment by vadim-moz** [comment]
@myronmarston : Thank you very much! We discussed with @koos42 , and decided to keep accessors for now -- we might switch / deprecate later. Your suggestions about methods names and signatures are great. Give us a moment!

File: spec/unit/plines/job_batch_spec.rb (hunk 1 above)
Comment (vadim-moz) [inline spec/unit/plines/job_batch_spec.rb]: Done.

**Comment by vadim-moz** [comment]
methods names & signatures fixed. PTAL

**Comment by myronmarston** [comment]
LGTM

**Comment by koos42** [comment]
Cool! Thanks!

**Pull request merged**